File: node/src/test/integration.test.ts
/**
 * Integration test: Send a message via REST API and receive it via Mercury WebSocket.
 *
 * This test verifies the entire pipeline:
 * 1. Device registration (WDM)
 * 2. Mercury WebSocket connection
 * 3. KMS initialization (ECDH handshake)
 * 4. Message send (REST API)
 * 5. Message receive (Mercury)
 * 6. Message decryption (KMS)
 *
 * Run with: WEBEX_BOT_TOKEN=your_token npm run test:integration
 */

import { WebexMessageHandler } from '../index.js';

const TIMEOUT_MS = 30000; // 30 seconds

interface WebexPerson {
  id: string;
  emails: string[];
  displayName: string;
}

async function integrationTest(): Promise<void> {
  const receiverToken = process.env.WEBEX_BOT_TOKEN;
  if (!receiverToken) {
    throw new Error('WEBEX_BOT_TOKEN environment variable is required (bot that receives messages)');
  }

  const senderToken = process.env.WEBEX_BOT_TOKEN_TEST;
  if (!senderToken) {
    throw new Error('WEBEX_BOT_TOKEN_TEST environment variable is required (bot that sends test message)');
  }

  console.log('🚀 Starting integration test...\n');

  // Create handler with receiver bot
  const handler = new WebexMessageHandler({ token: receiverToken });

  // Unique test message
  const testMessage = `Integration test ${Date.now()}`;
  let receivedMessage = false;
  let receivedText = '';

  handler.on('message:created', (msg) => {
    console.log(`📨 Received message: "${msg.text}" from ${msg.personEmail}`);
    if (msg.text === testMessage) {
      receivedMessage = true;
      receivedText = msg.text;
    }
  });

  handler.on('connected', () => {
    console.log('✅ Connected to Mercury');
  });

  handler.on('error', (err) => {
    console.error('❌ Handler error:', err);
  });

  try {
    // Step 1: Connect to Mercury
    console.log('1️⃣  Connecting to Mercury...');
    await handler.connect();

    // Step 2: Get receiver bot's email
    console.log('2️⃣  Fetching bot identities...');
    const receiverResponse = await fetch('https://webexapis.com/v1/people/me', {
      headers: { Authorization: `Bearer ${receiverToken}` }
    });
    const senderResponse = await fetch('https://webexapis.com/v1/people/me', {
      headers: { Authorization: `Bearer ${senderToken}` }
    });

    if (!receiverResponse.ok || !senderResponse.ok) {
      throw new Error('Failed to get bot identities');
    }

    const receiver = await receiverResponse.json() as WebexPerson;
    const sender = await senderResponse.json() as WebexPerson;
    console.log(`   Receiver: ${receiver.displayName} (${receiver.emails[0]})`);
    console.log(`   Sender: ${sender.displayName} (${sender.emails[0]})`);

    // Step 3: Send message FROM sender bot TO receiver bot
    console.log(`3️⃣  Sending test message: "${testMessage}"`);
    const sendResponse = await fetch('https://webexapis.com/v1/messages', {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${senderToken}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        toPersonEmail: receiver.emails[0],
        text: testMessage
      })
    });

    if (!sendResponse.ok) {
      const errorBody = await sendResponse.text();
      throw new Error(`Failed to send message: ${sendResponse.status} ${errorBody}`);
    }

    const sentMessage = await sendResponse.json() as { id: string };
    console.log(`   Message sent (ID: ${sentMessage.id})`);

    // Step 4: Wait for message to arrive via Mercury
    console.log('4️⃣  Waiting for message to arrive via Mercury...');
    const startTime = Date.now();
    while (!receivedMessage && (Date.now() - startTime) < TIMEOUT_MS) {
      await new Promise(resolve => setTimeout(resolve, 500));
    }

    // Step 5: Verify result
    console.log('\n📊 Test Results:');
    if (receivedMessage) {
      console.log('✅ PASSED - Message received and decrypted successfully');
      console.log(`   Expected: "${testMessage}"`);
      console.log(`   Received: "${receivedText}"`);
    } else {
      console.error('❌ FAILED - Message not received within timeout');
      throw new Error('Integration test failed: message not received');
    }

  } finally {
    // Cleanup
    console.log('\n🧹 Cleaning up...');
    await handler.disconnect();
    console.log('✅ Disconnected\n');
  }
}

// Run the test
integrationTest()
  .then(() => {
    console.log('✅ Integration test completed successfully');
    process.exit(0);
  })
  .catch((error) => {
    console.error('❌ Integration test failed:', error);
    process.exit(1);
  });

File: go/integration_test.go
package webexmessagehandler_test

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"testing"
	"time"

	webex "github.com/3rg0n/webex-message-handler/go"
)

// Integration test: Send a message via REST API and receive it via Mercury WebSocket.
//
// This test verifies the entire pipeline:
// 1. Device registration (WDM)
// 2. Mercury WebSocket connection
// 3. KMS initialization (ECDH handshake)
// 4. Message send (REST API)
// 5. Message receive (Mercury)
// 6. Message decryption (KMS)
//
// Run with: WEBEX_BOT_TOKEN=receiver_token WEBEX_BOT_TOKEN_TEST=sender_token go test -v -run TestIntegration

const timeoutSeconds = 30

type webexPerson struct {
	ID          string   `json:"id"`
	Emails      []string `json:"emails"`
	DisplayName string   `json:"displayName"`
}

type webexMessage struct {
	ID string `json:"id"`
}

func TestIntegrationSendAndReceive(t *testing.T) {
	receiverToken := os.Getenv("WEBEX_BOT_TOKEN")
	if receiverToken == "" {
		t.Skip("WEBEX_BOT_TOKEN environment variable not set (bot that receives messages)")
	}

	senderToken := os.Getenv("WEBEX_BOT_TOKEN_TEST")
	if senderToken == "" {
		t.Skip("WEBEX_BOT_TOKEN_TEST environment variable not set (bot that sends test message)")
	}

	fmt.Println("\n🚀 Starting integration test...")

	// Create handler with receiver bot
	handler, err := webex.New(webex.Config{
		Token: receiverToken,
	})
	if err != nil {
		t.Fatalf("Failed to create handler: %v", err)
	}

	// Unique test message
	testMessage := fmt.Sprintf("Integration test %d", time.Now().UnixMilli())
	receivedChan := make(chan string, 1)

	handler.OnMessageCreated(func(msg webex.DecryptedMessage) {
		fmt.Printf("📨 Received message: \"%s\" from %s\n", msg.Text, msg.PersonEmail)
		if msg.Text == testMessage {
			select {
			case receivedChan <- msg.Text:
			default:
			}
		}
	})

	handler.OnConnected(func() {
		fmt.Println("✅ Connected to Mercury")
	})

	handler.OnError(func(err error) {
		fmt.Printf("❌ Handler error: %v\n", err)
	})

	ctx := context.Background()

	// Step 1: Connect to Mercury
	fmt.Println("1️⃣  Connecting to Mercury...")
	if err := handler.Connect(ctx); err != nil {
		t.Fatalf("Failed to connect: %v", err)
	}
	defer handler.Disconnect(ctx)

	// Step 2: Get both bot identities
	fmt.Println("2️⃣  Fetching bot identities...")
	receiver, err := getWhoAmI(receiverToken)
	if err != nil {
		t.Fatalf("Failed to get receiver bot identity: %v", err)
	}
	sender, err := getWhoAmI(senderToken)
	if err != nil {
		t.Fatalf("Failed to get sender bot identity: %v", err)
	}
	fmt.Printf("   Receiver: %s (%s)\n", receiver.DisplayName, receiver.Emails[0])
	fmt.Printf("   Sender: %s (%s)\n", sender.DisplayName, sender.Emails[0])

	// Step 3: Send message FROM sender bot TO receiver bot
	fmt.Printf("3️⃣  Sending test message: \"%s\"\n", testMessage)
	msgID, err := sendMessage(senderToken, receiver.Emails[0], testMessage)
	if err != nil {
		t.Fatalf("Failed to send message: %v", err)
	}
	fmt.Printf("   Message sent (ID: %s)\n", msgID)

	// Step 4: Wait for message to arrive via Mercury
	fmt.Println("4️⃣  Waiting for message to arrive via Mercury...")
	select {
	case receivedText := <-receivedChan:
		// Step 5: Verify result
		fmt.Println("\n📊 Test Results:")
		fmt.Println("✅ PASSED - Message received and decrypted successfully")
		fmt.Printf("   Expected: \"%s\"\n", testMessage)
		fmt.Printf("   Received: \"%s\"\n", receivedText)
		if receivedText != testMessage {
			t.Errorf("Message mismatch: expected %q, got %q", testMessage, receivedText)
		}
	case <-time.After(timeoutSeconds * time.Second):
		fmt.Println("\n📊 Test Results:")
		fmt.Println("❌ FAILED - Message not received within timeout")
		t.Fatal("Integration test failed: message not received")
	}

	// Cleanup
	fmt.Println("\n🧹 Cleaning up...")
	handler.Disconnect(ctx)
	fmt.Println("✅ Disconnected")
}

func getWhoAmI(token string) (*webexPerson, error) {
	req, err := http.NewRequest("GET", "https://webexapis.com/v1/people/me", nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, body)
	}

	var person webexPerson
	if err := json.NewDecoder(resp.Body).Decode(&person); err != nil {
		return nil, err
	}

	return &person, nil
}

func sendMessage(token, toEmail, text string) (string, error) {
	body := map[string]string{
		"toPersonEmail": toEmail,
		"text":          text,
	}
	bodyJSON, _ := json.Marshal(body)

	req, err := http.NewRequest("POST", "https://webexapis.com/v1/messages", strings.NewReader(string(bodyJSON)))
	if err != nil {
		return "", err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("HTTP %d: %s", resp.StatusCode, body)
	}

	var msg webexMessage
	if err := json.NewDecoder(resp.Body).Decode(&msg); err != nil {
		return "", err
	}

	return msg.ID, nil
}
